File: src/harness/adapters/claude_code.py
"""claude-code adapter — invokes the `claude` CLI in print mode.

Output format: --output-format json gives a structured envelope:
    { "type": "result", "result": "...", "usage": {...}, "total_cost_usd": ... }

Session transcripts live at ``<config>/projects/<encoded cwd>/<session-id>.jsonl``.
The layout below was read from the Claude Code 2.1.220 native binary's embedded
bundle (OpenClaude, a fork, ships the same algorithm in
``src/utils/sessionStoragePortable.ts``):

- config root: ``NFC(CLAUDE_CONFIG_DIR ?? ~/.claude)``
- project key: ``NFC(realpath(cwd))``, or ``NFC(cwd)`` when realpath fails
- directory name: every char outside ``[a-zA-Z0-9]`` becomes ``-``; names longer
  than 200 chars are cut to 200 and suffixed ``-<base36(abs(djb2(project key)))>``
- lookups also accept sibling directories sharing that 200-char prefix

JavaScript works on UTF-16 code units, so the sanitizer and hash do too.
"""
from __future__ import annotations

import json
import os
import re
import string
import struct
import unicodedata
from pathlib import Path
from typing import Iterable

from harness._subproc import SubprocOutcome
from harness.base import (
    Adapter,
    AgentStatus,
    BuildCommand,
    InstallMeta,
    ParsedOutput,
    ReadyState,
    RunSpec,
    ScrollKeys,
    SessionTelemetry,
)
from harness.pricing import derive_cost
from harness.util import last_non_empty_join, strip_ansi


_CLAUDE_CODE_FULLSCREEN_SCROLL_KEYS = ScrollKeys(line_down="C-M-e", line_up="C-M-y", page_down="NPage", page_up="PPage")

_PROMPT_LINE_RE = re.compile(r"^\s*[>❯]\s*$")
_BYPASS_RE = re.compile(r"bypass.?permissions", re.IGNORECASE)
_ACCEPT_RE = re.compile(r"Yes, I accept", re.IGNORECASE)
_TRUST_RE = re.compile(r"trust this folder|Do you trust the files", re.IGNORECASE)
_UPDATE_RE = re.compile(r"Update available", re.IGNORECASE)
_RATE_LIMIT_RE = re.compile(r"rate.?limit|hit your limit", re.IGNORECASE)
_TIMER_RE = re.compile(r"\((?:\d+m\s+)?\d+s[\s·)]")
_STATUS_BAR_RE = re.compile(r"bypass permissions|Claude Code", re.IGNORECASE)

PROJECT_DIR_NAME_LIMIT = 200
_PROJECT_DIR_SAFE_UNITS = frozenset(map(ord, string.ascii_letters + string.digits))
_BASE36_DIGITS = string.digits + string.ascii_lowercase


def _utf16_units(text: str) -> tuple[int, ...]:
    raw = text.encode("utf-16-le", "surrogatepass")
    return struct.unpack(f"<{len(raw) // 2}H", raw)


def _djb2(units: Iterable[int]) -> int:
    """Upstream ``((hash << 5) - hash + charCode) | 0`` over UTF-16 code units."""
    h = 0
    for unit in units:
        h = ((h << 5) - h + unit) & 0xFFFFFFFF
    return h - 0x100000000 if h >= 0x80000000 else h


def _base36(n: int) -> str:
    if n == 0:
        return "0"
    digits = []
    while n:
        n, rem = divmod(n, 36)
        digits.append(_BASE36_DIGITS[rem])
    return "".join(reversed(digits))


def encode_project_path(project_path: str) -> str:
    """Directory name Claude-family CLIs use for ``project_path`` under ``projects/``."""
    units = _utf16_units(project_path)
    sanitized = "".join(chr(u) if u in _PROJECT_DIR_SAFE_UNITS else "-" for u in units)
    if len(sanitized) <= PROJECT_DIR_NAME_LIMIT:
        return sanitized
    return f"{sanitized[:PROJECT_DIR_NAME_LIMIT]}-{_base36(abs(_djb2(units)))}"


def canonical_project_path(workdir: Path) -> str:
    """``NFC(realpath(workdir))``; upstream keeps the unresolved path when realpath fails."""
    absolute = os.path.abspath(workdir)
    try:
        resolved = os.path.realpath(absolute, strict=True)
    except OSError:
        resolved = absolute
    return unicodedata.normalize("NFC", resolved)


def config_home(env_var: str, default_dirname: str, *, empty_is_unset: bool = False) -> Path:
    """NFC config root; Claude preserves an explicit empty path, OpenClaude does not."""
    explicit = os.environ.get(env_var)
    root = explicit if explicit is not None and (explicit or not empty_is_unset) else os.path.join(os.path.expanduser("~"), default_dirname)
    return Path(unicodedata.normalize("NFC", root)).absolute()


def project_dirs(projects_root: Path, project_path: str) -> list[Path]:
    """Existing transcript directories for ``project_path``, exact match first.

    Mirrors upstream resume lookups: a truncated name also matches sibling
    directories that share its 200-char prefix (hash suffixes differ across
    runtimes and versions).
    """
    encoded = encode_project_path(project_path)
    exact = projects_root / encoded
    dirs = [exact] if exact.is_dir() else []
    if len(encoded) <= PROJECT_DIR_NAME_LIMIT:
        return dirs
    prefix = encoded[:PROJECT_DIR_NAME_LIMIT] + "-"
    try:
        siblings = sorted(
            p for p in projects_root.iterdir() if p.name.startswith(prefix) and p.is_dir() and not p.is_symlink()
        )
    except OSError:
        return dirs
    dirs.extend(p for p in siblings if p != exact)
    return dirs


def _transcript_matches_project(path: Path, project_path: str) -> bool:
    # Bounded metadata lookup: path sanitization and truncated prefixes collide.
    # Missing identity is unavailable, never permission to select a foreign log.
    with path.open("rb") as stream:
        header = stream.read(65536).decode("utf-8", errors="replace")
    for line in header.splitlines():
        try:
            event = json.loads(line)
        except json.JSONDecodeError:
            continue
        if not isinstance(event, dict) or event.get("isSidechain"):
            continue
        cwd = event.get("cwd")
        if isinstance(cwd, str) and cwd and os.path.isabs(cwd):
            return canonical_project_path(Path(cwd)) == project_path
    return False


def newest_session_log(dirs: Iterable[Path], project_path: str, session_started_after: float | None) -> str | None:
    """Newest matching-project JSONL at/after the epoch-seconds mtime cutoff."""
    newest: tuple[float, Path] | None = None
    for d in dirs:
        try:
            entries = list(d.iterdir())
        except OSError:
            continue
        for p in entries:
            try:
                if p.suffix != ".jsonl" or not p.is_file():
                    continue
                mtime = p.stat().st_mtime
                if session_started_after is not None and mtime < session_started_after:
                    continue
                if newest is not None and (mtime, str(p)) <= (newest[0], str(newest[1])):
                    continue
                if _transcript_matches_project(p, project_path):
                    newest = (mtime, p)
            except OSError:
                continue
    return str(newest[1]) if newest else None


def _token_count(value: object) -> int | None:
    """Match JavaScript's safe-integer range instead of accepting lossy counts."""
    if value is None:
        return 0
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not 0 <= value <= 2**53 - 1:
        return None
    count = int(value)
    return count if count == value else None


def parse_claude_transcript(path: str) -> SessionTelemetry:
    """Sum assistant usage once per API message id; derive cost when no cost field exists.

    Claude Code writes one transcript line per assistant content block, each
    repeating the same ``message.id`` and ``message.usage``; upstream groups
    those lines by id, so usage is counted once per id here.
    """
    p = Path(path)
    if not p.exists():
        return SessionTelemetry(path, None, None, None, None, None)

    tokens_in = 0
    tokens_out = 0
    cost_usd = 0.0
    model_name: str | None = None
    saw_usage = False
    saw_cost = False
    seen_message_ids: set[str] = set()
    previous_usage: dict[str, tuple[int, int]] = {}

    try:
        for line in p.read_text(encoding="utf-8").splitlines():
            t = line.strip()
            if not t:
                continue
            try:
                event = json.loads(t)
            except json.JSONDecodeError:
                continue
            if not isinstance(event, dict):
                continue
            msg = event.get("message") if isinstance(event.get("message"), dict) else {}
            usage = msg.get("usage") if isinstance(msg.get("usage"), dict) else {}
            message_id = msg.get("id")
            duplicate = isinstance(message_id, str) and message_id in seen_message_ids
            if isinstance(message_id, str):
                seen_message_ids.add(message_id)
            if usage:
                current_in = _token_count(usage.get("input_tokens"))
                current_out = _token_count(usage.get("output_tokens"))
                if current_in is None or current_out is None:
                    return SessionTelemetry(path, None, None, None, None, None)
                prior_in, prior_out = previous_usage.get(message_id, (0, 0)) if isinstance(message_id, str) else (0, 0)
                tokens_in += current_in - prior_in
                tokens_out += current_out - prior_out
                if tokens_in > 2**53 - 1 or tokens_out > 2**53 - 1:
                    return SessionTelemetry(path, None, None, None, None, None)
                if isinstance(message_id, str):
                    previous_usage[message_id] = (current_in, current_out)
                saw_usage = True
            event_cost = event.get("costUSD")
            if not isinstance(event_cost, (int, float)) or isinstance(event_cost, bool):
                event_cost = event.get("total_cost_usd")
            if isinstance(event_cost, (int, float)) and not isinstance(event_cost, bool) and not duplicate:
                saw_cost = True
                cost_usd += float(event_cost)
            # Skip claude-code's "<synthetic>" placeholder — an autoresponder /
            # interrupt turn, not a real model invocation.
            if model_name is None and isinstance(msg.get("model"), str) and msg.get("model") != "<synthetic>":
                model_name = msg.get("model")
    except OSError:
        return SessionTelemetry(path, None, None, None, None, None)

    final_in = tokens_in if saw_usage else None
    final_out = tokens_out if saw_usage else None
    final_cost = cost_usd if saw_cost else derive_cost(model_name, final_in, final_out)
    return SessionTelemetry(path, final_in, final_out, final_cost, model_name, None)


def _has_prompt_line(pane: str) -> bool:
    return any(_PROMPT_LINE_RE.match(line.strip()) for line in strip_ansi(pane).split("\n"))


class ClaudeCodeAdapter(Adapter):
    name = "claude-code"
    instructions_filename = "CLAUDE.md"
    scroll_ownership = "fullscreen-aware"
    permission_bypass_args = ("--dangerously-skip-permissions",)
    native_options_kind = "claude-code"
    config_home_env = "CLAUDE_CONFIG_DIR"
    config_file_flag = "--settings"
    submit_keys = ("Enter",)
    install_meta = InstallMeta(
        package_manager="npm",
        install_command=("npm", "install", "-g", "@anthropic-ai/claude-code"),
        update_command=("npm", "install", "-g", "@anthropic-ai/claude-code@latest"),
        version_command=("claude", "--version"),
    )

    DEFAULT_MODEL = "sonnet"

    def get_current_scroll_keys(self) -> ScrollKeys | None:
        # claude-code's `/tui` slash command toggles between the classic
        # main-screen renderer ("default") and the alt-screen virtualized
        # renderer ("fullscreen"). Persisted under `tui` in the user-scope
        # settings.json inside the config root. v1: user-scope only — full
        # managed → local → project → user precedence is intentionally
        # deferred.
        return _CLAUDE_CODE_FULLSCREEN_SCROLL_KEYS if _read_claude_code_tui_mode() == "fullscreen" else None

    def build_command(self, spec: RunSpec) -> BuildCommand:
        resolved = self.resolve_run_spec(spec)
        args = [
            "-p", spec.prompt,
            "--model", resolved.model,
            *resolved.native_args,
            "--output-format", "json",
            *resolved.permission_args,
            *resolved.config_args,
        ]
        # -p mode does not auto-walk workdir for CLAUDE.md; inject explicitly so
        # the instructions are always visible to the model.
        if spec.instructions:
            args += ["--append-system-prompt", spec.instructions]
        return self.finalize_command(spec, cmd="claude", args=args)

    def parse_output(self, spec: RunSpec, outcome: SubprocOutcome) -> ParsedOutput:
        raw: dict | None = None
        if outcome.stdout.strip():
            try:
                raw = json.loads(outcome.stdout)
            except json.JSONDecodeError:
                raw = None

        cost = tokens_in = tokens_out = None
        if isinstance(raw, dict):
            usage = raw.get("usage") or {}
            tokens_in = usage.get("input_tokens")
            tokens_out = usage.get("output_tokens")
            cost = raw.get("total_cost_usd")

        return {"cost_usd": cost, "tokens_in": tokens_in, "tokens_out": tokens_out, "raw": raw}

    # ---- session-aware ---------------------------------------------------

    def detect_ready(self, pane: str) -> ReadyState:
        last20 = last_non_empty_join(pane, 20)
        if _has_prompt_line(pane) and _STATUS_BAR_RE.search(strip_ansi(pane)):
            return "ready"
        if _BYPASS_RE.search(last20) and _ACCEPT_RE.search(last20):
            return "dialog"
        if _TRUST_RE.search(last20):
            return "dialog"
        if _UPDATE_RE.search(last20) and "?" in last20:
            return "dialog"
        return "loading"

    def handle_dialog(self, pane: str) -> list[str] | None:
        text = strip_ansi(pane)
        if _BYPASS_RE.search(text) and _ACCEPT_RE.search(text):
            return ["2", "Enter"]
        if _TRUST_RE.search(text):
            return ["Enter"]
        # Decline updates mid-run; an explicit update command is the install path.
        if _UPDATE_RE.search(text):
            return ["Escape"]
        return None

    def detect_status(self, pane: str) -> AgentStatus:
        last10 = last_non_empty_join(pane, 10)
        if _RATE_LIMIT_RE.search(last10):
            return "rate-limited"
        if _UPDATE_RE.search(last10) and "?" in last10:
            return "dialog"
        # claude-code shows "(Xs · ↑M ↓N)" or "·X tokens·" while running
        if _TIMER_RE.search(last10):
            return "running"
        # Idle: prompt visible without timer
        if _has_prompt_line(pane):
            return "idle"
        return "unknown"

    def session_log_path(self, workdir: Path, session_started_after: float | None = None) -> str | None:
        projects = config_home(self.config_home_env, ".claude") / "projects"
        project_path = canonical_project_path(workdir)
        return newest_session_log(project_dirs(projects, project_path), project_path, session_started_after)

    def parse_session_log(self, path: str) -> SessionTelemetry:
        return parse_claude_transcript(path)


def _read_claude_code_tui_mode() -> str | None:
    settings_path = config_home("CLAUDE_CONFIG_DIR", ".claude") / "settings.json"
    if not settings_path.exists():
        return None
    try:
        raw = json.loads(settings_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    if isinstance(raw, dict):
        v = raw.get("tui")
        return v if isinstance(v, str) else None
    return None

File: ts/src/adapters/claude-code.ts
import { register } from '../registry.js'
import type { Adapter, AgentStatus, BuildCommand, ParsedOutput, ReadyState, RunSpec, ScrollKeys, SessionTelemetry, SubprocOutcome } from '../base.js'
import { finalizeCommand, validateRunSpec } from '../base.js'
import { stripAnsi, lastNonEmptyJoin } from '../util.js'
import { deriveCost } from '../pricing.js'
import { closeSync, existsSync, openSync, readFileSync, readSync, readdirSync, realpathSync, statSync } from 'node:fs'
import { homedir } from 'node:os'
import { join, resolve } from 'node:path'

// Session transcripts live at <config>/projects/<encoded cwd>/<session-id>.jsonl.
// Read from the Claude Code 2.1.220 native binary's embedded bundle (OpenClaude,
// a fork, ships the same algorithm in src/utils/sessionStoragePortable.ts):
//   config root:    NFC(CLAUDE_CONFIG_DIR ?? ~/.claude)
//   project key:    NFC(realpath(cwd)), or NFC(cwd) when realpath fails
//   directory name: /[^a-zA-Z0-9]/g -> '-'; names over 200 chars are cut to 200
//                   and suffixed '-' + base36(abs(djb2(project key)))
//   lookups also accept sibling directories sharing that 200-char prefix

export const PROJECT_DIR_NAME_LIMIT = 200

function djb2(text: string): number {
  let hash = 0
  for (let i = 0; i < text.length; i++) hash = ((hash << 5) - hash + text.charCodeAt(i)) | 0
  return hash
}

/** Directory name Claude-family CLIs use for `projectPath` under `projects/`. */
export function encodeProjectPath(projectPath: string): string {
  const sanitized = projectPath.replace(/[^a-zA-Z0-9]/g, '-')
  if (sanitized.length <= PROJECT_DIR_NAME_LIMIT) return sanitized
  return `${sanitized.slice(0, PROJECT_DIR_NAME_LIMIT)}-${Math.abs(djb2(projectPath)).toString(36)}`
}

/** NFC(realpath(workdir)); upstream keeps the unresolved path when realpath fails. */
export function canonicalProjectPath(workdir: string): string {
  const absolute = resolve(workdir)
  let resolved = absolute
  try { resolved = realpathSync(absolute) } catch { /* keep the unresolved path, like upstream */ }
  return resolved.normalize('NFC')
}

/**
 * NFC config root; Claude preserves an explicit empty path, OpenClaude does not.
 * Node's os.homedir() reads $HOME per call but Bun caches it at startup, so
 * $HOME is consulted directly first.
 */
export function configHome(envVar: string, defaultDirname: string, emptyIsUnset = false): string {
  const explicit = process.env[envVar]
  return resolve((explicit !== undefined && (explicit || !emptyIsUnset)
    ? explicit : join(process.env['HOME'] || homedir(), defaultDirname)).normalize('NFC'))
}

/**
 * Existing transcript directories for `projectPath`, exact match first. Mirrors
 * upstream resume lookups: a truncated name also matches sibling directories
 * sharing its 200-char prefix (hash suffixes differ across runtimes/versions).
 */
export function projectDirs(projectsRoot: string, projectPath: string): string[] {
  const encoded = encodeProjectPath(projectPath)
  const exact = join(projectsRoot, encoded)
  const dirs: string[] = []
  try { if (statSync(exact).isDirectory()) dirs.push(exact) } catch { /* no exact directory */ }
  if (encoded.length <= PROJECT_DIR_NAME_LIMIT) return dirs
  const prefix = encoded.slice(0, PROJECT_DIR_NAME_LIMIT) + '-'
  try {
    const siblings = readdirSync(projectsRoot, { withFileTypes: true })
      .filter(d => d.isDirectory() && d.name.startsWith(prefix))
      .map(d => join(projectsRoot, d.name))
      .sort()
    for (const candidate of siblings) if (candidate !== exact) dirs.push(candidate)
  } catch { /* projects root unreadable: exact match only */ }
  return dirs
}

function transcriptMatchesProject(path: string, projectPath: string): boolean {
  const fd = openSync(path, 'r')
  let header: string
  try {
    const buffer = Buffer.allocUnsafe(65536)
    header = buffer.toString('utf8', 0, readSync(fd, buffer, 0, buffer.length, 0))
  } finally { closeSync(fd) }
  for (const line of header.split('\n')) {
    let event: unknown
    try { event = JSON.parse(line) } catch { continue }
    const record = asRecord(event)
    if (!record || record['isSidechain']) continue
    const cwd = record['cwd']
    if (typeof cwd === 'string' && cwd.startsWith('/')) return canonicalProjectPath(cwd) === projectPath
  }
  return false
}

/** Newest matching-project JSONL at/after the epoch-milliseconds mtime cutoff. */
export function newestSessionLog(dirs: string[], projectPath: string, since?: number): string | null {
  let newest: { mtimeMs: number; path: string } | null = null
  for (const dir of dirs) {
    let names: string[]
    try { names = readdirSync(dir) } catch { continue }
    for (const name of names) {
      if (!name.endsWith('.jsonl')) continue
      const path = join(dir, name)
      let stat
      try { stat = statSync(path) } catch { continue }
      if (!stat.isFile()) continue
      if (since !== undefined && stat.mtimeMs < since) continue
      if (newest !== null && (stat.mtimeMs < newest.mtimeMs || (stat.mtimeMs === newest.mtimeMs && path <= newest.path))) continue
      try {
        if (transcriptMatchesProject(path, projectPath)) newest = { mtimeMs: stat.mtimeMs, path }
      } catch { /* Unreadable artifact or missing identity. */ }
    }
  }
  return newest?.path ?? null
}

function asRecord(value: unknown): Record<string, unknown> | undefined {
  return value !== null && typeof value === 'object' && !Array.isArray(value) ? value as Record<string, unknown> : undefined
}

/**
 * Sum assistant usage once per API message id; derive cost when no cost field
 * exists. Claude Code writes one transcript line per assistant content block,
 * each repeating the same `message.id` and `message.usage`; upstream groups
 * those lines by id, so usage is counted once per id here.
 */
export function parseClaudeTranscript(path: string): SessionTelemetry {
  if (!existsSync(path)) {
    return { sessionLogPath: path, tokensIn: null, tokensOut: null, costUsd: null, model: null, raw: null }
  }
  let tokensIn = 0, tokensOut = 0, costUsd = 0, modelName: string | null = null
  let sawUsage = false, sawCost = false
  const seenMessageIds = new Set<string>()
  const previousUsage = new Map<string, { input: number; output: number }>()
  try {
    for (const line of readFileSync(path, 'utf-8').split('\n')) {
      const t = line.trim()
      if (!t) continue
      let ev: unknown
      try { ev = JSON.parse(t) } catch { continue }
      const obj = asRecord(ev)
      if (!obj) continue
      const msg = asRecord(obj['message'])
      const usage = asRecord(msg?.['usage'])
      const messageId = msg?.['id']
      const duplicate = typeof messageId === 'string' && seenMessageIds.has(messageId)
      if (typeof messageId === 'string') seenMessageIds.add(messageId)
      if (usage) {
        const input = Number(usage['input_tokens'] ?? 0)
        const output = Number(usage['output_tokens'] ?? 0)
        const prior = typeof messageId === 'string' ? previousUsage.get(messageId) : undefined
        tokensIn += input - (prior?.input ?? 0)
        tokensOut += output - (prior?.output ?? 0)
        if (typeof messageId === 'string') previousUsage.set(messageId, { input, output })
        sawUsage = true
      }
      const cost = typeof obj['costUSD'] === 'number' ? obj['costUSD'] : obj['total_cost_usd']
      if (typeof cost === 'number' && !duplicate) { sawCost = true; costUsd += cost }
      const model = msg?.['model']
      // Skip claude-code's "<synthetic>" placeholder — that's an
      // autoresponder/interrupt turn, not a real model invocation. Pick the
      // first NON-synthetic string instead.
      if (typeof model === 'string' && model !== '<synthetic>' && !modelName) modelName = model
    }
  } catch {
    return { sessionLogPath: path, tokensIn: null, tokensOut: null, costUsd: null, model: null, raw: null }
  }
  const finalTokensIn = sawUsage ? tokensIn : null
  const finalTokensOut = sawUsage ? tokensOut : null
  const finalCost = sawCost ? costUsd : deriveCost(modelName, finalTokensIn, finalTokensOut)
  return { sessionLogPath: path, tokensIn: finalTokensIn, tokensOut: finalTokensOut, costUsd: finalCost, model: modelName, raw: null }
}

const claudeCodeAdapter: Adapter = {
  name: 'claude-code',
  instructionsFilename: 'CLAUDE.md',
  defaultModel: 'sonnet',
  scrollOwnership: 'fullscreen-aware',
  permissionBypassArgs: ['--dangerously-skip-permissions'],
  nativeOptionsKind: 'claude-code',
  configHomeEnv: 'CLAUDE_CONFIG_DIR',
  configFileFlag: '--settings',

  buildCommand(spec: RunSpec): BuildCommand {
    const validated = validateRunSpec(this, spec)
    const { model, permissionArgs, nativeArgs, configArgs } = validated
    const args = ['-p', spec.prompt, '--model', model, ...nativeArgs, '--output-format', 'json', ...permissionArgs, ...configArgs]
    // -p mode does not auto-walk workdir for CLAUDE.md; inject explicitly so
    // the instructions are always visible to the model.
    if (spec.instructions) {
      args.push('--append-system-prompt', spec.instructions)
    }
    return finalizeCommand(this, spec, validated, { cmd: 'claude', args })
  },

  parseOutput(_spec: RunSpec, outcome: SubprocOutcome): ParsedOutput {
    let raw: unknown = null
    if (outcome.stdout.trim()) {
      try {
        raw = JSON.parse(outcome.stdout)
      } catch {
        raw = null
      }
    }
    if (raw !== null && typeof raw === 'object' && raw !== null) {
      const obj = raw as Record<string, unknown>
      const usage = (obj['usage'] as Record<string, unknown> | undefined) ?? {}
      return {
        costUsd: (obj['total_cost_usd'] as number | undefined) ?? null,
        tokensIn: (usage['input_tokens'] as number | undefined) ?? null,
        tokensOut: (usage['output_tokens'] as number | undefined) ?? null,
        raw,
      }
    }
    return { costUsd: null, tokensIn: null, tokensOut: null, raw }
  },
}

// ===== session-aware additions =====

claudeCodeAdapter.submitKeys = ['Enter']

// claude-code's `/tui` slash command toggles between the classic main-screen
// renderer ("default") and the alt-screen virtualized renderer ("fullscreen").
// Persisted under the `tui` key in the user-scope settings.json inside the
// config root. When fullscreen, we forward C-M-e/y + NPage/PPage into the app;
// when default (or absent), real terminal scrollback works and we fall through
// to tmux.
//
// v1: user-scope only. The full claude-code precedence ladder
// (managed → local → project → user) is intentionally deferred — user-scope
// is where /tui-style global preferences land and is sufficient in practice.
const CLAUDE_CODE_FULLSCREEN_SCROLL_KEYS: ScrollKeys = {
  lineDown: 'C-M-e',
  lineUp: 'C-M-y',
  pageDown: 'NPage',
  pageUp: 'PPage',
}

function readClaudeCodeTuiMode(): string | null {
  const path = join(configHome('CLAUDE_CONFIG_DIR', '.claude'), 'settings.json')
  if (!existsSync(path)) return null
  try {
    const raw = JSON.parse(readFileSync(path, 'utf-8')) as unknown
    if (raw && typeof raw === 'object') {
      const v = (raw as Record<string, unknown>)['tui']
      return typeof v === 'string' ? v : null
    }
  } catch {
    // unreadable / malformed settings → treat as "default"
  }
  return null
}

claudeCodeAdapter.getCurrentScrollKeys = function (): ScrollKeys | null {
  return readClaudeCodeTuiMode() === 'fullscreen' ? CLAUDE_CODE_FULLSCREEN_SCROLL_KEYS : null
}

claudeCodeAdapter.detectReady = function (pane: string) {
  const last20 = lastNonEmptyJoin(pane, 20)
  const all = stripAnsi(pane)
  const hasPrompt = stripAnsi(pane).split('\n').some(l => /^\s*[>❯]\s*$/.test(l.trim()))
  const hasStatusBar = /bypass permissions/i.test(all) || /Claude Code/i.test(all)
  if (hasPrompt && hasStatusBar) return 'ready'
  if (/bypass.?permissions/i.test(last20) && /Yes, I accept/i.test(last20)) return 'dialog'
  if (/trust this folder/i.test(last20) || /Do you trust the files/i.test(last20)) return 'dialog'
  if (/Update available/i.test(last20) && /\?/.test(last20)) return 'dialog'
  return 'loading'
}

claudeCodeAdapter.handleDialog = function (pane: string) {
  const text = stripAnsi(pane)
  if (/bypass.?permissions/i.test(text) && /Yes, I accept/i.test(text)) return ['2', 'Enter']
  if (/trust this folder/i.test(text) || /Do you trust the files/i.test(text)) return ['Enter']
  // Decline updates mid-run; explicit `flt clis update` is the install path.
  if (/Update available/i.test(text)) return ['Escape']
  return null
}

claudeCodeAdapter.detectStatus = function (pane: string) {
  const last10 = lastNonEmptyJoin(pane, 10)
  if (/rate.?limit|hit your limit/i.test(last10)) return 'rate-limited'
  if (/Update available/i.test(last10) && /\?/.test(last10)) return 'dialog'
  // claude-code shows "(Xs · ↑M ↓N)" or "·X tokens·" while running
  if (/\((?:\d+m\s+)?\d+s[\s·)]/.test(last10)) return 'running'
  // Idle: prompt visible without timer
  if (stripAnsi(pane).split('\n').some(l => /^\s*[>❯]\s*$/.test(l.trim()))) return 'idle'
  return 'unknown'
}

claudeCodeAdapter.sessionLogPath = function (workdir: string, since?: number): string | null {
  const projects = join(configHome('CLAUDE_CONFIG_DIR', '.claude'), 'projects')
  const projectPath = canonicalProjectPath(workdir)
  return newestSessionLog(projectDirs(projects, projectPath), projectPath, since)
}

claudeCodeAdapter.parseSessionLog = parseClaudeTranscript

claudeCodeAdapter.installMeta = {
  packageManager: 'npm',
  installCommand: ['npm', 'install', '-g', '@anthropic-ai/claude-code'],
  updateCommand: ['npm', 'install', '-g', '@anthropic-ai/claude-code@latest'],
  versionCommand: ['claude', '--version'],
}

register('claude-code', claudeCodeAdapter)
